source/s3_client.c

```diff
@@ -1252,26 +1167,122 @@ static void s_s3_client_vip_connection_on_acquire_request_connection(
-    }
 
-    /* Notify the caller of their HTTP connection. */
-    if (payload->callback != NULL) {
-        payload->callback(*current_http_connection, AWS_ERROR_SUCCESS, payload->user_data);
+        AWS_LOGF_INFO(
+            AWS_LS_S3_CLIENT,
+            "id=%p Incoming connection has changed on VIP Connection %p.  Resetting local request count.",
+            (void *)client,
+            (void *)vip_connection);
+    } else {
+        ++vip_connection->request_count;
+
+        AWS_LOGF_INFO(
+            AWS_LS_S3_CLIENT,
+            "id=%p Incoming connection has NOT changed on VIP Connection %p.  Increasing local request count. %d",
+            (void *)client,
+            (void *)vip_connection,
+            vip_connection->request_count);
     }
 
+    aws_s3_meta_request_make_request(meta_request, client, vip_connection);
+
 clean_up:
 
-    aws_s3_client_release(payload->client);
-    aws_mem_release(client->allocator, payload);
-    payload = NULL;
+    s_s3_client_internal_release(client);
+}
+
+/* Called by the meta request when it has finished using this VIP connection for a single request. */
+void aws_s3_client_notify_connection_finished(
+    struct aws_s3_client *client,
+    struct aws_s3_vip_connection *vip_connection) {
+    AWS_PRECONDITION(client);
+    AWS_PRECONDITION(vip_connection);
+
+    /* If this VIP connection has a meta request, then it's a finished request, so clear it out. */
+    if (vip_connection->request != NULL) {
+        aws_s3_request_release(vip_connection->request);
+        vip_connection->request = NULL;
+    }
+
+    s_s3_client_lock_synced_data(client);
+
+    aws_linked_list_push_back(&client->synced_data.pending_vip_connection_updates, &vip_connection->node);
+    s_s3_client_schedule_process_work_task_synced(client);
+
+    s_s3_client_unlock_synced_data(client);
+}
+
+void aws_s3_client_notify_request_destroyed(struct aws_s3_client *client) {
+    s_s3_client_lock_synced_data(client);
+    ++client->synced_data.pending_request_count;
+    s_s3_client_schedule_process_work_task_synced(client);
+    s_s3_client_unlock_synced_data(client);
+}
+
+struct s3_streaming_body_payload {
+    struct aws_s3_client *client;
+    struct aws_linked_list requests;
+    struct aws_task task;
+};
+
+void aws_s3_client_stream_response_body(
+    struct aws_s3_client *client,
+    struct aws_s3_meta_request *meta_request,
+    struct aws_linked_list *requests) {
+    AWS_PRECONDITION(client);
+    AWS_PRECONDITION(requests);
+
+    struct s3_streaming_body_payload *payload =
+        aws_mem_calloc(client->sba_allocator, 1, sizeof(struct s3_streaming_body_payload));
+
+    s_s3_client_internal_acquire(client);
+    payload->client = client;
+
+    aws_linked_list_init(&payload->requests);
+    aws_linked_list_move_all_back(&payload->requests, requests);
+
+    aws_task_init(&payload->task, s_s3_client_body_streaming_task, payload, "s3_client_body_streaming_task");
```

Comment:
Oh this is interesting, this is going to be the bottleneck on PUT most likely. Looks like you've already moved it off the main I/O event loop, so that might be all the mitigation we can do right now. Let's keep an eye on this when we test puts.

Reply:
The body streaming stuff is only for gets.  We do the stream read for uploads in the request prepare.  (I was looking at doing the stream read in a separate thread, but walked it back when NUMA concerns were flagged.)